sdk/lib/opentelemetry/sdk/internal.rb

```diff
@@ -16,14 +16,39 @@ def boolean?(value)
         value.is_a?(TrueClass) || value.is_a?(FalseClass)
       end
 
+      def homogenous_element_types?(value)
+        first_element_type = nil
+        value.each do |element|
+          element_type = case element
+                         when String then String
+                         when TrueClass, FalseClass then :boolean
+                         when Numeric then Numeric
+                         end
+          if !first_element_type
+            first_element_type = element_type
+          elsif element_type != first_element_type
+            return false
+          end
+        end
+        true
+      end
+
       def valid_key?(key)
         key.instance_of?(String)
       end
 
-      def valid_value?(value)
+      def valid_primitive_value?(value)
         value.instance_of?(String) || value == false || value == true || value.is_a?(Numeric)
       end
 
+      def valid_array_value?(value)
+        value.is_a?(Array) && value.all? { |elt| valid_primitive_value?(elt) } && homogenous_element_types?(value)
+      end
+
+      def valid_value?(value)
+        valid_primitive_value?(value) || valid_array_value?(value)
+      end
+
```

Comment:
And then write this as:

```ruby
      def valid_value?(value)
        value.instance_of?(String) || value == false || value == true || value.is_a?(Numeric) || valid_array_value?(value)
      end

      def valid_array_value?(value)
        return false unless value.is_a?(Array)

        case value.first
        when String
          value.all? { |v| v.instance_of?(String) }
        when TrueClass, FalseClass
          value.all? { |v| boolean?(v) }
        when Numeric
          value.all? { |v| v.is_a?(Numeric) }
        else
          false
        end
      end
```

WDYT?

Reply:
Probably also need to handle the empty array case.

Reply:
Good call.

Reply:
I could be convinced otherwise on this, but I think there's some value to having standalone predicates named `valid_primitive_value?` and `homogenous_element_types?`. `valid_primitive_value?` gets used twice on my original code, where the suggested code buries the second check amid the `case` statement in `valid_array_value?`. And it seems more readable to highlight `homogenous_element_types?` as a specific, distinct condition of validity for array values.

Reply:
@mwear any thoughts on my comment from Sunday?

Reply:
@fhwang I strongly prefer the code that I suggested both for its simplicity and its performance. I've provided a benchmark for each implementation for both string and numeric arrays. The best situation for your implementation is the string case, where it is 2.58x slower than my suggestion. The worse case for your implementation is the numeric case, where it is 8.21x slower than my suggestion. The reason there is such a difference in performance between the two implementations is because my suggestion does a a few upfront checks, but a single comparison for each array element. Your implementation does as many as 10 comparisons per array element depending on the type.

### Benchmark Results

The benchmark results are below. My implementation is "valid array value 1", your is "valid array value 2":

```
Warming up --------------------------------------
valid array value 1 string
                        97.090k i/100ms
valid array value 2 string
                        44.723k i/100ms
Calculating -------------------------------------
valid array value 1 string
                          1.608M (± 2.3%) i/s -      8.058M
valid array value 2 string
                        624.411k (± 3.5%) i/s -      3.131M

Comparison:
valid array value 1 string:  1608102.0 i/s
valid array value 2 string:   624411.4 i/s - 2.58x slower

Warming up --------------------------------------
valid array value 1 numeric
                        91.511k i/100ms
valid array value 2 numeric
                        15.725k i/100ms
Calculating -------------------------------------
valid array value 1 numeric
                          1.369M (± 1.4%) i/s -      6.863M
valid array value 2 numeric
                        166.807k (± 1.8%) i/s -    849.150k

Comparison:
valid array value 1 numeric:  1368667.9 i/s
valid array value 2 numeric:   166807.2 i/s - 8.21x slower
```

The Benchmark:

```ruby
require 'benchmark/ips'
require 'securerandom'

string_array = Array.new(10) { SecureRandom.hex(8) }
numeric_array = Array.new(10) { rand(10) }

# my suggestion
def valid_array_value1?(value)
  return false unless value.is_a?(Array)
  return true if value.empty?

  case value.first
  when String
    value.all? { |v| v.instance_of?(String) }
  when TrueClass, FalseClass
    value.all? { |v| boolean?(v) }
  when Numeric
    value.all? { |v| v.is_a?(Numeric) }
  else
    false
  end
end

# original code
def homogenous_element_types?(value)
  first_element_type = nil
  value.each do |element|
    element_type = case element
                   when String then String
                   when TrueClass, FalseClass then :boolean
                   when Numeric then Numeric
                   end
    if !first_element_type
      first_element_type = element_type
    elsif element_type != first_element_type
      return false
    end
  end
  true
end

def valid_primitive_value?(value)
  value.instance_of?(String) || value == false || value == true || value.is_a?(Numeric)
end

def valid_array_value2?(value)
  value.is_a?(Array) && value.all? { |elt| valid_primitive_value?(elt) } && homogenous_element_types?(value)
end

Benchmark.ips do |x|
  x.report('valid array value 1 string') do
    valid_array_value1?(string_array)
  end
  x.report('valid array value 2 string') do
    valid_array_value2?(string_array)
  end
  x.compare!
end

Benchmark.ips do |x|
  x.report('valid array value 1 numeric') do
    valid_array_value1?(numeric_array)
  end
  x.report('valid array value 2 numeric') do
    valid_array_value2?(numeric_array)
  end
  x.compare!
end
```

Reply:
@mwear Thanks for digging into this. It definitely seems like your suggested change is faster. I actually think it wasn't just calling `valid_primitive_value?` on all elements, because when I restricted that call to just the first it was still noticeably slower. At any rate I've implemented the change you suggested.